Dynamic controller quorum (#203)
strimzi/proposals: 132-dynamic-controller-quorum.md

* delete the controllers' `StrimziPodSet`(s) so that the controller pods are deleted.
* update the number of replicas within the `KafkaNodePool`(s) custom resource(s) related to controllers (increase or decrease).
* unpause the cluster reconciliation.

The Strimzi Cluster Operator will be able to create the new controllers from scratch (with the updated number of replicas) and it will restart the brokers with the new controller quorum configured via the `controller.quorum.voters` property.
Of course, right after the controllers deletion step and until the controller quorum runs again, the cluster won't be available because brokers cannot work without the KRaft controllers running.

This is not parity features with ZooKeeper-based cluster where the ZooKeeper ensemble can be scaled up or down without cluster downtime.

## Motivation

KRaft dynamic controller quorum enables the controllers scaling without downtime.
In general, this operation can be useful for increasing the number of controllers when needed (or, of course, decreasing them).
Also, sometimes, replacing a controller because of a disk or hardware failure is something that can be done by scaling.

> Furthermore, even migration from an architecture with combined-nodes to controller dedicated nodes could be achieved by scaling the KRaft quorum.

Reviewer: ```suggestion
It can also support migration from combined nodes to dedicated controller nodes by scaling the KRaft quorum.
```

It's being set when the controller storage is properly formatted and the quorum is configured as dynamic or static.
The only case needing the feature to be changed is when migrating from static to dynamic quorum (so updating the value from `0` to `1`) for an existing Apache Kafka cluster, when coming from an older operator release.

The following sections describes what's the idea behind the proposal and how to approach the above steps.
As a referece, [here](http://kafka.apache.org/41/operations/kraft/#provisioning-nodes)'s the link to the official Apache Kafka documentation about using dynamic quorum.

### Understanding Static vs Dynamic Quorum

The main difference between static and dynamic quorum lies in where the voters set membership information is stored and how it can be changed.

With **static quorum** (`kraft.version=0`), the quorum membership is hardcoded in the node configuratrion file by using the `controller.quorum.voters` property listing all the controllers which take part at the KRaft quorum.
This membership configuration never changes during the cluster's lifetime and exists only in each node's configuration file.

> The voters set is immutable and there is no `VotersRecord` stored in the metadata log.

Reviewer: ```suggestion
The voter set is immutable and there is no `VotersRecord` stored in the metadata log.
```

### New cluster creation and reconciliation with dynamic quorum

The initial cluster creation is going to use the [bootstrap with multiple controllers](https://kafka.apache.org/41/operations/kraft/#bootstrap-with-multiple-controllers) approach.
The Strimzi Cluster Operator:

* builds the broker and controller configuration by setting the `controller.quorum.bootstrap.servers` field (instead of the `controller.quorum.voters` one).
* generates a random directory ID, as a UUID, for each controller.
* saves the `controllers` field (list of `KafkaControllerStatus` objects) within the `Kafka` custom resource status, containing the controller IDs and their corresponding directory IDs.
* builds the controllers string from the controllers status list, and stores it as `controllers` field within the node ConfigMap, to be loaded by the `kafka_run.sh` script where it's needed for formatting the storage properly.

On node start up, the `kafka_run.sh` script loads the `controllers` field from the node ConfigMap into a `CONTROLLERS` variable.

> Since all clusters use dynamic quorum (either from creation or after automatic migration), this variable is always set.

Reviewer: ```suggestion
Since all clusters use dynamic quorum (either from creation or after automatic migration logic is applied), this variable is always set.
```

When a running controller is shut down, it has to be unregistered in order to leave the KRaft quorum.
The KRaft quorum reconciliation is a critical process that ensures the KRaft controller quorum membership accurately reflects the desired state of the cluster.
The main goal is reconciling the current KRaft quorum gathered from the Apache Kafka cluster with the desired controllers state requested by the users.
This process handles controller registration and unregistration needed for scaling up/down operations, metadata disk changes and disaster recovery.
It also takes part within the migration from static to dynamic quorum process.
The proposal is to have a new dedicated `KRaftQuorumReconciler` class to manage the necessary reconciliation steps.

The KRaft quorum reconciliation fits into the `KafkaReconciler` reconciliation process in this way:

* BEGIN RECONCILIATION
  * ... (other reconcile operations) ...
  * **KRaft quorum reconciliation**: analyze current quorum state, unregister and register controllers as needed (typically unregisters controllers being scaled down).
  * scale down controllers.

>   * ... (other reconcile operations) ...

Author: @tinaselenge thanks for the feedback, following the rationale and story behind this approach ...

At the beginning of prototyping this, I had two different methods: `unregisterControllers` to come before scale down and `registerControllers` to come after scale up. The I decided to consolidate into a single reconcile KRaft quorum because they looked quite similar in terms of logic but ... the thing is that it still remains the fact that unregistration has to be done before scaling down (otherwise you can brake the quorum if scaling down before unregistration), and registration has to be done after scale up (because the controllers don't exist yet if you think to register them beforehand).
Now, if we want to have just one place to reconcile I see:

* at the beginning: it works out of box for scaling down (unregistration first) but not for scale up. It means that when you are scaling up controllers in one reconciliation, the registration will happen later on in a subsequent reconciliation. It would work, as you mentioned it's idempotent. But it will happen with some delay (depending on the reconciliation period, 2 mins by default). Unless some state changed in the Kafka CR and a new reconciliation will be triggered immediately.
* at the end: it works out of box for scaling up (registration of newly added controllers) but it doesn't for scale down. As mentioned, you cannot scale down without unregistering controllers first. To have quorum reconciliation at the end only, we should add a logic to block the scale down if there are still registered controllers (something similat to what we have to block scale down if brokers are hosting partitions). I can have a think about it.

>   * ... (other reconcile operations) ...

Author: @tinaselenge I thoughts about the idea of having the KRaft quorum reconciliation just at the end, but it can't work easily in case of disk metadata change. Let's consider this scenario:

- The `KafkaRoller` detects there is a disk change (It has to unregister old controller and register the new one)
- It runs the unregistration first but ... it FAILS
- The overall operator reconciliation fails

Now, with the current implementation:
  - On next reconciliation, at the beginning, the call to `KRaftQuorumReconciler` retries to unregister the failed one, then registering the new one
  - WARN: this is needed because otherwise the KafkaRoller can't proceed with the next controller to be rolled

But with `KRaftQuorumReconciler` at the end only:
  - On next reconciliation, at the beginning, there is NO call to `KRaftQuorumReconciler` and the flow goes directly to `KafkaRoller` without unregistering the failed one.
  - The `KafkaRoller` will just start to go through the other controllers to be rolled but without unregistering the already rolled one first.

NOTE: the same scenario could happen with the unregistration working but the registration failing on first reconciliation.

When there is a metadata disk change we have following constrained flow:

For each controller in the list:
- roll the controller
- controller starts and new metadata disk is formatted (it's needed for reading the meta.properties during KRaft Quorum reconciliation)
- KRaft Quorum reconciliation on single node (unregister old incarnation, register new incarnation)

We CAN'T proceed with the new controller to be rolled until the previous one is gone through the above flow.

# Support for dynamic controller quorum and controllers scaling

This proposal introduces support for KRaft dynamic controller quorums in the Strimzi Cluster Operator to enable controller scaling.
The dynamic controller quorum was introduced in Kafka 3.9.0 through [KIP-853](https://cwiki.apache.org/confluence/display/KAFKA/KIP-853%3A+KRaft+Controller+Membership+Changes).

## Current situation

Currently, the Strimzi Cluster Operator supports only the KRaft static controller quorum, even though Apache Kafka 4.x supports dynamic quorum.
The static controller quorum doesn't allow scaling the controller nodes without downtime.

> It means that after the Apache Kafka cluster creation, when the controller nodes are deployed, it is not possible to add or remove controllers to/from the quorum anymore, unless accepting some downtime for the Apache Kafka cluster.

Reviewer 2: I think there is a downtime of the controller quorum. But not of the brokers. It is similar to when ZooKeeper was unavailable, or?

Author: Of course, all operations which involves metadata changes won't work with the controller quorum being down (i.e. admin on topics, users, ...). In some situations, even producers and consumers could not work if they need a metadata refresh (i.e. a partition leader goes down and needs to be re-elected, ...).

Reviewer 2: So it is not necessarily a downtime. _Availability might be limited_ during that time.

Author: I will change the wording.

A possible way for scaling controllers is:

* pause the cluster reconciliation.
* delete the controllers' `StrimziPodSet`(s) so that the controller pods are deleted.
* update the number of replicas within the `KafkaNodePool`(s) custom resource(s) related to controllers (increase or decrease).

> * unpause the cluster reconciliation.

Reviewer 2: I don't think this is needed. You can just change it and once it gets stuck rolling the controllers you can just roll the remaining nodes manually.

Author: Yeah, I was mentioning a "possible way" but even what you are describing should do the same without the need to pause/unpause. I will change it if you like, no strong opinion. It's going to be just an example of how it (doesn't) work today.

Reviewer 2: Actually, scale-up seems to work fine. Scale-down require manual deletion of one or more pods.

It might be also worth mentioning the node unregistration issues discussed today on Slack as that is IMHO much bigger issue here if it cannot be worked around.

Author: > It might be also worth mentioning the node unregistration issues discussed today on Slack as that is IMHO much bigger issue here if it cannot be worked around.

The discussion was about a cluster using static quorum and a user trying to scale controllers which is officially not supported in Apache Kafka. Controllers registration/unregistration doesn't apply when you have static quorum so I can't understand what you are referring to.

Reviewer 2: You are talking here about what the _static quorum cannot do_. So my comment is that you make some parts of it make much more complicated then they really are but at the same time you are skipping some important parts.

The Strimzi Cluster Operator will be able to create the new controllers from scratch (with the updated number of replicas) and it will restart the brokers with the new controller quorum configured via the `controller.quorum.voters` property.
Of course, right after the controllers deletion step and until the controller quorum runs again, the cluster won't be available because brokers cannot work without the KRaft controllers running.

This does not provide feature parity with a ZooKeeper-based cluster, where the ZooKeeper ensemble can scale up or down without cluster downtime.

> ## Motivation

Reviewer 2: Are there any plans to remove the static quorum in Kafka? I think it would be useful to note here.

Author: To be honest I couldn't find anything official about that (there is no mention within KIP-853 either). But I see "signals" about its deprecation and I guess future removal. For example, all the properties files provided by the Apache Kafka distribution are defaulting to use the `controller.quorum.bootstrap.servers` and also Kafka 4.2 is defaulting to `kraft.version` be set as 1 which means using dynamic quorum. Also within the Kafka documentation, regarding the `controller.quorum.voters` parameter, [here](https://kafka.apache.org/42/configuration/broker-configs/#brokerconfigs_controller.quorum.voters), you can see how it's referred as "the old way of defining membership for controller quorums". But I will let @showuon @mimaison @fvaleri to provide any official source/plan they are aware of (if any).

Reviewer 3: AFAIK, for new Kafka 4.0+ deployments, dynamic quorum is recommended. Static quorum remains supported but is the legacy path.

This is from KIP-853 discussion thread on dev mailing list, which makes me think we may want to keep it as a fallback mechanism.

> The precedence order is the VotersRecord in the snapshot and log. If
there are no VotersRecords in the snapshot or log, it will use the
configuration in controller.quorum.voters.

KRaft dynamic controller quorum enables the controllers scaling without downtime.
In general, this operation can be useful for increasing the number of controllers when needed (or, of course, decreasing them).

> Also, sometimes, replacing a controller because of a disk or hardware failure is something that can be done by scaling.

Reviewer 2: I do not think you need scaling when you have broken disk or bare-metal server:
* You can easily just fix the existing controller (e.g. by having it scheduled on a new node or start with a new disk)
* Solving this by scaling does nto make sense as you would have the broken node missing (of you scale-down and remove it) or still there breaking your reconciliation. And it will screw up your qourum.

Reviewer 4: The KIP seemed to say that to replace the disk of a controller node you would need to load some state onto the disk, so my understanding from the KIP is that if you have a hardware failure the dynamic quorum would allow you to remove that node and add a new one with a fresh disk without having to load that state somehow, is that right @ppatierno ?

Author: It's true that a disk failure can be easily fixed by adding a new disk and restarting the controller. The KIP-853 explains it [here](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=217391519#KIP853:KRaftControllerMembershipChanges-Diskfailurerecovery) how. @katheris, maybe this is the part you are referring to.
But it's also true that you can scale up your quorum with a new controller and then scale down the failed one (actually removing it from the quorum). Of course, it seems too much compared to just replace the disk. But it's feasible. Anyway, I agree and will update the sentence by not mentioning the scaling but mentioning that the disk replacement works because the dynamic quorum with the corresponding unregistration/registration controller operations.

Reviewer 2: So, the point is ...
* Disk replacement works fine with static quorum because you just replace it I guess.
* Disk replacement might work fine with a dynamic quorum, but we should not force users to scale up and down for it. The cloud native way is to fix the disk in the existing Pod. So how will that be handled in this proposal? I assume it is more similar to the JBOD storage handling and to moving metadata between volumes?

Author: I removed the sentence about scaling for fixing disk failures.

> So how will that be handled in this proposal? I assume it is more similar to the JBOD storage handling and to moving metadata between volumes?

I think it's already explained across the proposal. For example, the user updates the `KafkaNodePool` by adding a new JBOD disk and marking it as the one to host metadata (user also removes the old disk). The change will trigger a controller rolling (as it happens already today) handled by the `KafkaRoller` and it will be in charge of unregistering the controller with the old directory ID and registering the same controller with the new directory ID.

> It can also support migration from combined nodes to dedicated controller nodes by scaling the KRaft quorum.

Reviewer 2: ```suggestion
It can also support migration from combined nodes to dedicated controller nodes by creating new controller-only nodes and removing the controller role from the existing mixed nodes.
```

Finally, this would provide feature parity with a ZooKeeper-based cluster, where ZooKeeper nodes can be added to or removed from the quorum.

Beyond these operational benefits, dynamic quorum support is critical for Apache Kafka's strategic direction and production readiness.
With ZooKeeper support officially removed in Apache Kafka 4.0, KRaft is now the only metadata management option for Kafka clusters.
This makes dynamic quorum capabilities essential rather than optional and it's a fundamental requirement for KRaft to be truly production-ready and not a regression from ZooKeeper-based deployments.

> However, it's important to note that while the core dynamic quorum functionality was introduced in Kafka 3.9.0 (via KIP-853), certain bug fixes, improvements and new features like quorum migration from static to dynamic are only available starting from Apache Kafka 4.1.0.

Reviewer 2: Are you saying that Strimzi was not production ready since 0.46 on? Are you saying that Kafka was not production ready at least in 3.9 and 4.0 with Kraft when it did not worked properly?

Author: I am just saying that Kafka 3.9 added the support for dynamic quorum but that version didn't have the support for migration from static to dynamic which was added in Kafka 4.0. This version also brought some improvements and bug fixes in the dynamic quorum itself.

Reviewer 2: You are saying that `dynamic quorum support is critical for Apache Kafka's production readiness`. So yes, I think you are saying that KRaft in Strimzi is not production ready. IIRC the dynamic quorum did not really work in practice 3.9.0 and 4.0.0. But to be clear, I do not think it is critical for production readiness. So it does not matter that much whether it worked or not.

* Node storage formatting is impacted by this change.
* The `controller.quorum.voters` property is now replaced by the `controller.quorum.bootstrap.servers` property on both broker and controller nodes.
* The `kraft.version` feature is now `1` when using dynamic quorum. It's `0` when using static quorum. 

By the way, regarding the last point, there is no need to set the `kraft.version` feature upfront (to `0` or `1`).
It's being set when the controller storage is properly formatted and the quorum is configured as dynamic or static.
The only case needing the feature to be changed is when migrating from static to dynamic quorum (so updating the value from `0` to `1`) for an existing Apache Kafka cluster, when coming from an older operator release.

The following sections describe the idea behind the proposal and how to approach the above steps.
For reference, see the [Apache Kafka documentation on provisioning nodes](http://kafka.apache.org/41/operations/kraft/#provisioning-nodes).

### Understanding Static vs Dynamic Quorum

The main difference between static and dynamic quorum lies in where the voter set membership information is stored and how it can be changed.

> With **static quorum** (`kraft.version=0`), the quorum membership is hardcoded in the node configuratrion file by using the `controller.quorum.voters` property listing all the controllers which take part at the KRaft quorum.

Reviewer 4: ```suggestion
With **static quorum** (`kraft.version=0`), the quorum membership is hardcoded in the node configuration file by using the `controller.quorum.voters` property listing all the controllers which take part at the KRaft quorum.
```

This membership configuration never changes during the cluster's lifetime and exists only in each node's configuration file.
The voter set is immutable and there is no `VotersRecord` stored in the metadata log.
When controllers start up, they read the voter sets directly from their local configuration and the leader discovery/election can start immediately.

With **dynamic quorum** (`kraft.version=1`), the quorum membership is stored in the replicated metadata log itself by using a `VotersRecord`.
The node configuration file uses the `controller.quorum.bootstrap.servers` property instead, which only provides contact points for initial discovery.
The voter set is mutable and can change through Raft consensus operations.
When controllers start up, they discover the voter sets by reading the `VotersRecord` from either the bootstrap snapshot (for initial controllers) or by fetching it from the leader metadata log (for newly added controllers, during a scale up).

The key differences can be summarized as:

> | Aspect | Static Quorum | Dynamic Quorum |

Reviewer 2: Pro-tip ... if you format this with spacing, it will be more easy to read during the review / as a raw file.

An alternative approach to bootstrapping a dynamic quorum cluster is to bootstrap a standalone controller first.
It means that, instead of formatting multiple controllers with the full controllers list, only a single controller is formatted and started initially using the `--standalone` flag.
This creates a quorum of one without requiring the controllers list but still writing a bootstrap snapshot file with a corresponding `VotersRecord`.
Once this standalone controller is running, additional controllers can be formatted with `--no-initial-controllers`, started as observers, and then dynamically added to the quorum using the standard add controller operations.
While this approach simplifies the initial bootstrap by avoiding the need to coordinate directory IDs for all controllers upfront, it means the cluster starts with no fault tolerance since a single-controller quorum cannot tolerate any failures.
However, once additional controllers are added and registered as voters, the cluster achieves the desired redundancy and fault tolerance.

> This approach can't cope with how Strimzi starts up the cluster nodes all together and doesn't have the possibility to do a rolling start one by one on cluster creation. 

Reviewer 4: I don't think this sentence is very clear, so suggesting the below to better explain the difficulties with this approach
```suggestion
Since the StrimziPodSet controller creates cluster nodes all together, not one-by-one, adopting this approach would require a fundamental change to the way we create the StrimziPodSets in the Kafka reconciliation.
```

This works when the Apache Kafka cluster is using the static quorum during a new cluster creation, when all nodes need to be formatted from scratch at the same time, as well as on cluster scaling, when a new broker node is added but the formatting is done the same way (of course controllers scaling is not supported).

Dynamic quorum changes how storage formatting works and makes a clear distinction on `kafka-storage.sh` usage when it's new cluster creation or node addition (broker or controller).
When the cluster is created, each broker can be formatted the same way as today, but the controller formatting needs the new `--initial-controllers` (or `-I`) option which specifies the controllers list.
The controllers list is a comma-separated list of controllers in the format `node-id@dns-name:port:directory-id`, representing the controllers that will form the KRaft quorum in the new cluster.

The formatting tool is then used the following way:

```shell
./bin/kafka-storage.sh format -t="$STRIMZI_CLUSTER_ID" -r="$METADATA_VERSION" -c=/tmp/strimzi.properties -g -I="$CONTROLLERS"
```

The `CONTROLLERS` variable is the one containing such a controllers list.

> The main challenge during cluster creation is differentiating between brokers and controllers to use the right options for the formatting.

Reviewer 2: Isn't the main challenge identifying the new cluster versus existing cluster? Distinguishing brokers and controllers seems simple. Or maybe you should explain why is it the main challenge.

Author: Yeah I will remove the sentence. In reality, you always know if a node is broker or controller (or both) by looking at the `process.roles` configuration parameter. What's really challenging is make the decision about how formatting a controller with `--initial-controllers` or `--no-initial-controllers` as explained later.

The broker storage formatting doesn't use the `--initial-controllers` option.

But when it comes to scaling up, and adding a new broker or a controller on an existing cluster, in both cases, the formatting tool uses the `--no-initial-controllers` (or `-N`) option instead. For controllers, this replaces `--initial-controllers`.

```shell
./bin/kafka-storage.sh format -t="$STRIMZI_CLUSTER_ID" -r="$METADATA_VERSION" -c=/tmp/strimzi.properties -g -N
```

So it means that when using the dynamic quorum, the script executed on node startup needs a way to differentiate between new cluster creation or node scaling in order to run the formatting with the appropriate options.

The above scenarios can be summarized this way:

* Static quorum:
  * Broker and controllers are formatted with no additional new options.
* Dynamic quorum:
  * New cluster creation:
    * broker doesn't need any additional options (but it works well with `--no-initial-controllers` as well to simplify the proposal, see later).
    * controller needs the controllers list via `--initial-controllers`.
  * Existing cluster, scale up:

>     * broker and controller needs the `--no-initial-controllers` option.

Reviewer 2: Why does new broker not need it and new broker in existing cluster does?

Author: This is a valid point. I went through it by testing and looking at the storage formatting tool in Kafka [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/tools/StorageTool.scala#L162). When you are formatting a broker, it seems that the flag doesn't matter at all. During my tests I was also able to scale up a new broker without such a flag. I will get confirmation about my discovery and will update the proposal accordingly.

> The proposal is for the Strimzi Cluster Operator to manage the list of current controllers, with their corresponding directory IDs, in the `Kafka` custom resource status.

Reviewer 4: This is a minor point, but in terms of addressing the requirement of passing the controllers list to the nodes it might make sense to mention the ConfigMap first, then mention that this data will also be stored in the status to be used for disaster recovery and for users to understand the set of registered voters more easily.

Author: The thing is that the content of the controllers list for the ConfigMap comes from the "controller" field within the status if it's not a new Kafka cluster. This is the reason why I am mentioning such status first.

This list is stored in a new `controllers` field which contains a list of `KafkaControllerStatus` objects, each having:

* `id`: The node ID of the controller.
* `directoryId`: The directory ID (UUID) of the controller used for KRaft quorum management.

For example:

```yaml
# ...
status:
  # ...
  controllers:
    - id: 3
      directoryId: "r9VGTiw2QUCoDUadCt6q2g"
    - id: 4
      directoryId: "r8E7BKRbR0SQxSIFro6wjg"
    - id: 5

>       directoryId: "Anjr8banTey_LqVZsppzYg"

Reviewer 2: We are getting significant push-back about this for the cluster-id. I think we need to listen to the users and avoid this.

Reviewer 4: I'm not aware of the push-back, are you able to expand or link to some discussion for me to see the context?

Reviewer 2: There are some issues and discussions. I think it was also two or three times raised at KubeCon. Basically, users have to get the cluster crashlooping to find the cluster ID, delete the cluster, set the ClusterID properly in the status and then recover the cluster. Now imagine doing this for all the metadata directories.

To be honest, I do not understand why the directory IDs are needed and matter to Kafka. Nobody managed to explain it to me. If we cannot get Kafka to get rid of them, we could also investigate the workarounds I suggested over a year ago which somehow did not even make it into the rejected alternatives. Such as designing the directory IDs ourself to make them fixed and predictable (i.e. combination of node ID and volume ID used to create the directory ID). The main question I never tested was whether we can easily override the existing directory IDs or not.

Author: > There are some issues and discussions. I think it was also two or three times raised at KubeCon. Basically, users have to get the cluster crashlooping to find the cluster ID, delete the cluster, set the ClusterID properly in the status and then recover the cluster. Now imagine doing this for all the metadata directories.

Let's consider that it's only in case of disaster recovery where I would expect users to do some manual steps. Said that, getting directory ids from the metadata directory is not something that can't be done automatically via a script (I have it for example). Also, how many metadata directories? 3 or 5? How many controllers do you expect an Apache Kafka cluster would have?

> To be honest, I do not understand why the directory IDs are needed and matter to Kafka. Nobody managed to explain it to me. 

Can you provide me reference about this discussions please? Any JIRAs or email on the Kafka mailing list? I couldn't find one maybe I am just bad at searching for JIRAs and emails.

>  If we cannot get Kafka to get rid of them, we could also investigate the workarounds I suggested over a year ago which somehow did not even make it into the rejected alternatives. Such as designing the directory IDs ourself to make them fixed and predictable (i.e. combination of node ID and volume ID used to create the directory ID).

Going through a deterministic approach was my first try. I had something like this in my code:

```java
String deterministicUuid = UUID.nameUUIDFromBytes( 
           (namespace + clusterName + nodeId).getBytes() 
).toString(); 
```

but it turned out this could not be enough given KIP-853 about [disk failure and replacement](https://cwiki.apache.org/confluence/display/KAFKA/KIP-853:+KRaft+Controller+Membership+Changes#KIP853:KRaftControllerMembershipChanges-Diskfailurerecovery) . The UUID should be unique per disk, not just per node. There is time window where the same controller "lives" with two directory IDs (old and new) and they need to be unregistered and registered as explained in this proposal. Maybe, for generating a UUID, we could attach the JBOD disk ID we have within the custom resource in Strimzi or (maybe better) the `metadata.log.dir` string. It has a format like this, `/var/lib/kafka/data-0/kafka-log3` (where 0 is the JBOD disk id and 3 is the nodeId, so maybe we could also avoid the nodeId as in the example).

Reviewer 2: > > If we cannot get Kafka to get rid of them, we could also investigate the workarounds I suggested over a year ago which somehow did not even make it into the rejected alternatives. Such as designing the directory IDs ourself to make them fixed and predictable (i.e. combination of node ID and volume ID used to create the directory ID).
> 
> Going through a deterministic approach was my first try. I had something like this in my code:
> 
> ```java
> String deterministicUuid = UUID.nameUUIDFromBytes( 
>            (namespace + clusterName + nodeId).getBytes() 
> ).toString(); 
> ```
> 
> but it turned out this could not be enough given KIP-853 about [disk failure and replacement](https://cwiki.apache.org/confluence/display/KAFKA/KIP-853:+KRaft+Controller+Membership+Changes#KIP853:KRaftControllerMembershipChanges-Diskfailurerecovery) . The UUID should be unique per disk, not just per node. There is time window where the same controller "lives" with two directory IDs (old and new) and they need to be unregistered and registered as explained in this proposal. Maybe, for generating a UUID, we could attach the JBOD disk ID we have within the custom resource in Strimzi or (maybe better) the `metadata.log.dir` string. It has a format like this, `/var/lib/kafka/data-0/kafka-log3` (where 0 is the JBOD disk id and 3 is the nodeId, so maybe we could also avoid the nodeId as in the example).

Yes, that is what I was proposing back then - have the directory UUID a combination of the node ID and volume ID (volume ID == the ID from the JBOD array). Paths come and go. So I would probably not base it on the exact path. But a string such as `node-1-data-0` is obviously fine.

Reviewer 3: Directory IDs make total sense in KIP-858 where they were first introduced. The directory ID identifies a physical disk, not a node. A new disk must get a new UUID so the controller knows it is empty and must not be assumed to contain the old partitions. This matters for partitions placement.

Instead, controllers have a single metadata log directory. There is no JBOD. There is no scenario where "disk 2 of 3 failed but the controller is still running." When a controller's disk fails, the controller is dead. In case of combined mode, I think the node would die too, but I have to try.

I can't find similar requirement in Raft literature. KIP-853 doesn't give any motivation either, so it may be something to raise with the upstream community.

* If the node is a controller:
    * Extracts the controller IDs from the `CONTROLLERS` string.
    * Checks if the current controller's ID is in the controllers list:
      * **Controller in list**:
        * If metadata disk change is detected, formats with `--no-initial-controllers` to generate a new directory ID (handles disk failures or storage replacement scenarios).
        * Otherwise, formats with `--initial-controllers` to preserve the directory ID specified in the controllers string. This handles initial bootstrap, rolling restarts, PVC replacements, and disaster recovery scenarios.
      * **Controller not in list**:
        * Formats with `--no-initial-controllers` to generate a new random directory ID (new controller being added during scale-up).

All the formatting operations also have the `-g` (so `--ignore-formatted`) flag to ensure already-formatted storage is not reformatted.

When a node has the combined role (both broker and controller), the script follows the controller formatting path since the `process.roles` includes being a controller.

### Reconcile an existing cluster with static quorum

> When the Strimzi Cluster Operator is upgraded to a version that supports dynamic quorum, it's going to reconcile existing `Kafka` custom resources for clusters which are using the static quorum.

Reviewer 2: So, does it reconcile it? Or do a migration? These seem to contradict each toher.

Author: For me "reconcile", as a verb, has the meaning of "taking care of it" so the operator always goes through a reconciliation loop, makes decisions and takes actions. In this case, the "reconcile" will end up with a "migration" operation from static to dynamic quorum as explained a few lines below. But if it's unclear to you I can change it. Any suggestion?

Reviewer 2: When you say that it is going to `reconcile` Kafka with static quorum, I take it that it will successfully pass the reconciliation loop with it. If it does `migration to dynamic quorum in the first reconciliation`, it should say so.

Reviewer 2: It is not just about the words. Lot of the things here seem to happen eventually. So there is a difference whethe rthe first reconciliation loop migrates it or whether it happens some reconciliations later.

Author: > When you say that it is going to reconcile Kafka with static quorum, I take it that it will successfully pass the reconciliation loop with it.

This is your take. As explaining I have a different way of reading the "reconcile" word. But I will make it clear for you.

Scaling controllers up or down within a KRaft quorum involves more than just spinning up new nodes or shutting down existing ones.
When a new controller is added to the Apache Kafka cluster, it has to be registered in order to join the KRaft quorum.
When a running controller is shut down, it has to be unregistered in order to leave the KRaft quorum.
The KRaft quorum reconciliation is a critical process that ensures the KRaft controller quorum membership accurately reflects the desired state of the cluster.
The main goal is to reconcile the current KRaft quorum state from the Apache Kafka cluster with the desired controller state requested by users.
This process handles controller registration and unregistration needed for scaling up/down operations, metadata disk changes and disaster recovery.
It also takes part within the migration from static to dynamic quorum process.
The proposal is to have a new dedicated `KRaftQuorumReconciler` class to manage the necessary reconciliation steps.

The KRaft quorum reconciliation fits into the `KafkaReconciler` reconciliation process in this way:

> * BEGIN RECONCILIATION

Reviewer 4: To be honest I don't find this part helpful, since it lists steps that don't exist explicitly in the existing reconciliation (I assume scale up actually just happens as part of the `podSet()` method) and doesn't explicitly state what the other reconcile operations are and where it fits amongst those. So I think either you should change it to have the specific method names in the existing code and call and the new methods, or just remove it. Also the `(rolling)` part doesn't make it clear that this would happen as part of the usual rolling update, that was something I only realised later in reading.

Author: I will update it with actual main methods that are called along the reconciliation mostly related to the points we are interested in (scale down, scale up, rolling + where the KRaft quorum reconciliation happens).

  * ... (other reconcile operations) ...
  * **KRaft quorum reconciliation**: analyze current quorum state, unregister and register controllers as needed (typically unregisters controllers being scaled down).
  * scale down controllers.
  * ... (other reconcile operations) ...
  * **KRaft quorum reconciliation** (rolling): for each controller pod restart, `KafkaRoller` invokes a "single-controller reconciliation" to handle metadata disk changes immediately (unregister old voter with stale directory ID, register new observer with current directory ID).
  * scale up new controllers.
  * ... (other reconcile operations) ...
  * **KRaft quorum reconciliation**: analyze current quorum state, unregister and register controllers as needed (typically registers newly added controllers).
  * ... (other reconcile operations) ...
* END RECONCILIATION

The following sections describe how KRaft quorum reconciliation works and how controller registration and unregistration are performed.

#### When it happens

> The KRaft quorum reconciliation is executed at two key points during the `KafkaReconciler` reconciliation process.

Reviewer 4: Here you have omitted any mention of calling the KRaft quorum reconciliation from the KafkaRoller which is confusing.

Author: I will add a sentence about it at the end of this section to clarify it's not in the `KafkaReconciler` but also within the `KafkaRoller`.

At the end of reconciliation, and after all rolling updates and scale up operations:
- Analyzes the quorum state again by querying the Kafka cluster.
- Performs unregistration and registration operations as needed. Typically registers newly added controllers that are running as observers and ready to join the quorum (i.e., caught up with the metadata log and synchronized with the active voters).

This two-phase approach ensures that:

- Controllers are unregistered before being deleted (during scale-down).
- Controllers are only registered after they have been rolled and are running with the correct configuration (during scale up or recovery).

This approach is similar to what is already used today for brokers unregistration which compares, at the end of each reconciliation, the list of registered brokers retrieved via the Kafka Admin API with the "desired" brokers running in the cluster, in order to determine which ones are missing and need to be unregistered.
In the controllers scaling scenario, the operator handles the registration as well (not only the unregistration).

> In addition to handling the registration and unregistration of controllers, the Strimzi Cluster Operator has to check the health of the KRaft quorum before allowing the scale down of the controllers.

Reviewer 2: So, what happens when the controllers are not ready?

Author: Not sure what do you mean by "not ready" but if for any reasons the operator can't get the health of the KRaft quorum, it should not allow to scale controllers but continuing the reconciliation. The Apache Kafka cluster would be still usable, from clients perspective (maybe admin operations not working because quourm not healthy?). Maybe I addressed more in the following comment https://github.com/strimzi/proposals/pull/203#discussion_r3016729246

Reviewer 2: > Not sure what do you mean by "not ready" but if for any reasons the operator can't get the health of the KRaft quorum, it should not allow to scale controllers but continuing the reconciliation.

I guess what if some of the pods are e.g. Pending or Crashlooping?

> The Apache Kafka cluster would be still usable, from clients perspective (maybe admin operations not working because quourm not healthy?). Maybe I addressed more in the following comment

Just a sidenote ... here the cluster without quorum is usable and basically ok. The same situation on the beginning with the static quorum is described as `downtime`. 😉

> At the beginning of the `KafkaReconciler` reconciliation process, if scaling down controllers would break the quorum (by losing the majority of voters needed for consensus), it should be blocked and reverted back by the operator.

Reviewer 2: What exactly does this mean?

Author: The idea is similar to what we have in `KafkaRoller` with `KafkaQuorumCheck`, which verifies quorum health before rolling a controller. The current check evaluate if there are enough controller caught up with the leader.
When scaling down controllers, we would need to check two things:                                                                                                                                                                       

* The remaining controllers still form a valid majority (i.e. 3 controllers -> need 2 for quorum)                                                                                                                   
* The remaining controllers are caught up with the leader

The check ensures that after removing controllers, you still have enough healthy controllers to maintain consensus.
For example if you want to scale from 5 to 3 controllers, the remaining would form a valid quorum but if 1 or 2 of those 3 are lagging behind in the metadata log, you'd only have 1-2 healthy controllers which are not enough for a majority of the 3-controller quorum. So the scale-down is blocked (like it happens when brokers are hosting partition and we block the scale down).

Reviewer 2: But Kafka Roller does not scale down things. So this would apply only when removing controller role from a mixed node I guess?

Author: > But Kafka Roller does not scale down things

Where do you see the Kafka Roller in the picture here?

> So this would apply only when removing controller role from a mixed node I guess?

Why? I am referring to scale down as well.

> A similar approach is already in place when attempting to scale down brokers that are hosting partitions, which could break HA.

Reviewer 4: So is the proposal to do this in the same way it's done for the brokers? So as part of the `KafkaClusterCreator` before the reconciliation starts, and then continuing with a reverted version of the CR in the internal state? If so that's not clear from the proposal atm in my opinion so it might be better to state it more explicitly.

Also this is a different kind of check I guess, since the broker one checks whether the broker has topic-partitions assigned to it which is something the user can change, whereas this is a check for something that is a more transient state so perhaps the operator should wait or retry at least once before taking that action.

Author: > So as part of the KafkaClusterCreator before the reconciliation starts, and then continuing with a reverted version of the CR in the internal state?

Correct. I will clarify.

> whereas this is a check for something that is a more transient state so perhaps the operator should wait or retry at least once before taking that action.

I am not sure how much it could be transient. It depends on in which situation is the KRaft quorum and why controller cannot be scaled down. In general, I prefer to avoid active waiting within a reconciliation (i.e. it doesn't happen for waiting rebalancing to finish) but I can understand the roller works differently to wait controllers readiness before proceeding.

#### How it works

The KRaft quorum reconciliation process follows these steps:

Query the current quorum state: Uses the Kafka Admin API `describeMetadataQuorum()` method to retrieve the current list of voters and observers, along with their directory IDs provided by the `QuorumInfo` object.

Analyze desired vs. actual state, for each "desired" controller node:
- Checks if the controller pod has been rolled with the controller role (verified by the presence of the `strimzi.io/controller-role` label on the pod).

> - Compares the controller's current state in the quorum (voter, observer, or absent) with the expected state based on the `controllers` status field.

Reviewer 4: This line confused me, because it seems like the status is not actually used for any checking during normal operation, I guess it's just used during a disaster recovery scenario? But that's not clear

Author: Yeah it is misleading, I will fix it. It's mostly described in this comment https://github.com/strimzi/proposals/pull/203#discussion_r3022580305

> - Determines if the controller needs to be registered, unregistered, or if it's already in the correct state.

Reviewer 2: I'm confused by this. What here represents the desired and actual state? It seems to be only actual state.

Author: by "state" I meant that ...
The describe quorum provides to the operator the source of truth and current state of the quorum, then the operator knows what are the desired controllers so it can compare and see what needs to be unregistered (because removed) and what needs to be registered (because added). I will make it clearer.

Detect and handle metadata disk changes, when a controller's disk is replaced (e.g., new JBOD disk, changing metadata disk, PVC replacement, disaster recovery), the controller will have a new directory ID:
- The quorum will show multiple incarnations of the same node ID. The old voter with the old directory ID, and a new observer with the new directory ID.
- The reconciliation process reads the actual directory ID from the controller pod's `meta.properties` file using the Kafka Agent.
- Unregisters the old voter with the stale directory ID.

> - Registers the new observer with the current directory ID.

Reviewer 2: * How does this deal with the current API and the related mechanism?
* Who / what does these things? (e.g. registers the observer etc.)

Author: > How does this deal with the current API and the related mechanism?

Which "current API and the related mechanism" are you referring to? Can clarify please?

> Who / what does these things? (e.g. registers the observer etc.)

It's explained in details in the next section "Reconciliation algorithms". It will be done by this new `KRaftQuorumReconciler` called by the `KafkaRoller` in this specific case of the disk metadata changes.

Execute quorum changes in two phases:
- Unregistration: Removes controllers from the quorum using the Kafka Admin API `removeRaftVoter()` method. This handles scale-down and removal of stale voters after metadata disk changes. Unregistration happens first to allow metadata disk changes scenarios where a node already exists as a voter with an old directory ID and needs to be replaced with a new one.
- Registration: Adds controllers to the quorum using the Kafka Admin API `addRaftVoter()` method. This handles scale-up and re-registration after metadata disk changes.

> Update the status: After successful reconciliation, the operator queries the quorum state one final time to retrieve the current voters and their directory IDs, then updates the `controllers` field in the `Kafka` custom resource status.

Reviewer 2: At which phase of the reconciliation is it updated?

Author: "After successful reconciliation" ... so at the end.

Retrieving the actual directory ID for a specific controller from the `meta.properties` file is a feature which is going to be added to the Kafka Agent and exposed through a dedicated `/directory-id` HTTP endpoint.
This endpoint accepts GET requests and returns the directory ID as a plain text UUID string read directly from the `meta.properties` file located in the metadata log directory.

> The Kafka Agent is also modified to receive the KRaft metadata log directory path via an environment variable, enabling it to locate and read the `meta.properties` file.

Reviewer 2: Why environment variable and not a parameter? We seem to use parameters right now.

Author: Not sure I wrote this. My POC is using a parameter passed via the `kafka-agent.properties` file as other parameters we already have there. I will fix the sentence.

> Importantly, this endpoint is only invoked when the reconciliation detects multiple incarnations of a controller (same node ID with different directory IDs), and is not called during normal operations, minimizing performance impact.

Reviewer 2: What exactly does this mean?

Author: In general, the describe metadata quorum through the Kafka Admin API always returns the directory IDs for the controllers but if you have multiple "incarnation" of a controller (because of disk changes) you can see two entries of the same controller with two different IDs and what's the actual one is retrieved by reading the `meta.properties` file on the node.

4. Execute changes via `executeQuorumChanges()` method in two phases:
   - Phase 1: Unregister all controllers in the `toUnregister` list (stale/unwanted controllers). If one of the controller is the leader, it will be unregistered as the last one to avoid useless leader elections in between multiple controllers unregistrations.
   - Phase 2: Register all controllers in the `toRegister` list (new controllers)
5. Read final quorum state using `describeMetadataQuorum()`
6. Build controller statuses from voters via `buildControllerStatuses()` method and return for the `Kafka` custom resource status update

**Single Controller Reconciliation** (called by KafkaRoller after individual pod restart):

1. Describe current metadata quorum state using `describeMetadataQuorum()`
2. Analyze only the restarted controller via `analyzeControllerNode()` method which builds `toRegister` and `toUnregister` lists (using same logic as full reconciliation)
3. Execute changes via `executeQuorumChanges()` method in two phases (unregister first, then register)
4. Done (status building happens in the full reconciliation)

#### Integration with rolling updates

> During rolling updates, the quorum reconciliation also operates at a finer granularity through the `KafkaRoller`:

Reviewer 4: I guess this is during any rolling update that is triggered for any reason (e.g. updating truststore, manually triggered, config change? But perhaps worth being explicit

Author: Yes, correct. I will make it explicit.

> - After each controller pod is restarted, the roller invokes a "single-controller reconciliation" to handle potential metadata disk changes immediately.

Reviewer 2: What exactly that is and what it means?

Author: It's explained in the "Immediate handling during rolling updates" section. If you read that, in summary, it's about reconciling the observer/voter status of the specific controller that was rolled and check if its old "incarnation" (with old directory ID) has to be unregistered and the new "incarnation" (with the new directory ID) has to be registered, before proceeding to roll the other controllers.

> - This ensures that if a controller's disk was replaced during the restart, the old voter is unregistered and the new one is registered right away, without waiting for the full reconciliation cycle.

Reviewer 2: How do you protect here from breaking the qourum?

Author: The `KafkaRoller` is already in charge of rolling the controller if changes are detected (including a disk change in the corresponding `KafkaNodePool`). It's also already in charge of doing that only if such a controller can be rolled without breaking the quorum, so the logic is already there today (taking this proposal apart). Only when it's rolled, and before proceeding to roll the next controller, there is the new logic of unregistering/registering the old/new "incarnation" of the controller. More details about how it works in terms of failures is in the section "Immediate handling during rolling updates" -> "Handling failures and recovery".

#### Handling failures

The KRaft quorum reconciliation is designed to be resilient to failures:

- If querying the quorum state fails, the reconciliation will fail and retry in the next reconciliation cycle.
- If reading a controller's directory ID fails (e.g., pod not ready, agent unavailable), the reconciliation fails for that controller and retries later.
- If unregistration or registration operations fail (e.g., controller not caught up, quorum not healthy), the operation is retried in the next reconciliation cycle.
- If the operator crashes in the mid-process, reconciliation runs again on start up.
- The process is idempotent: running it multiple times with the same desired state will converge to the correct result without causing issues.

This reconciliation approach ensures that the KRaft quorum membership is always kept in sync with the desired cluster state, handling scale-up, scale-down, metadata disk changes and recovery scenarios transparently.

### Adding new controllers (scale up)

When the controllers are scaled up, the Strimzi Cluster Operator:

> * builds the new controllers configuration by setting the `controller.quorum.bootstrap.servers` field.

Reviewer 2: How? Dynamically? Rolling update? Why is the bootstrap not a service?

Author: > How? Dynamically? Rolling update?

The `controller.quorum.bootstrap.servers` is defined as `READ_ONLY` within Kafka itself because it's needed at start up only. When it's updated, it means the corresponding node `ConfigMap` hosting the configuration is updated (so that it's reflected into the `/tmp/strimzi.properties` on the node) but the node doesn't need rolling as well as it's not updated dynamically. It's because the implicit nature of dynamic quorum, where nodes don't need such configuration to be reloaded (as it is for the "voters" configuration). They figure out controllers quorum via the `VotersRecord` and the metadata topic. So this config is used when a node is restarted for any other reasons at some point.

> Why is the bootstrap not a service?

I could investigate. Potentially it should work. This parameter is just for initial discovery so picking any controller as part of the quorum via a Service should be ok.

* starts up all the newly added controllers in parallel. Registration is handled later by reconciliation when the controllers are ready.

On node start up, within the `kafka_run.sh` script:

> * the controller node is formatted by using the `--no-initial-controllers` option.

Reviewer 2: How do you know how it should be used?

Author: The current controller node is not part of the "controllers" field loaded from the ConfigMap (as an env var) as explained in the "New cluster creation and reconciliation with dynamic quorum", where it mentions:

> **Controller not in list**:
        Formats with `--no-initial-controllers` to generate a new random directory ID (new controller being added during scale-up).

The Strimzi Cluster Operator will also update the `controller.quorum.bootstrap.servers` configuration within the ConfigMap(s) for the existing brokers and controllers.
It will not be used immediately because all nodes will get the information about newly added controllers dynamically via the KRaft protocol.
The new `controller.quorum.bootstrap.servers` configuration will be used on next start up, loaded from the updated ConfigMap(s).

When the controller is up and running, the Strimzi Cluster Operator:

* adds/registers the controllers within the quorum, one by one in sequence, by using the `addRaftVoter` method in the Kafka Admin API.

In order to add a controller, its directory ID is needed together with the corresponding endpoints (advertised listeners).

> The directory ID can be easily retrieved by using the Kafka Admin Client API `describeMetadataQuorum()` method, and extract the `ReplicaState.replicaDirectoryId()` for the specific controller (by using its ID).

Reviewer 2: Why dont' we just use the agent if we have it there?

Author: With just one `describeMetadataQuorum()` you get the directory IDs for all the controllers instead of several calls to the Kafka Agent to read the `meta.properties` node by node (where it's used to deal with two different "incarnation" of the same controller on disk changes, see comment here https://github.com/strimzi/proposals/pull/203#discussion_r3021813551)

But it also needs the advertised listeners for the new controller, via a `Set<RaftVoterEndpoint> endpoints` parameter.
The Kafka Admin Client API exposes such information via the `Node` class which is not provided by the `voters()` and `observers()` in the `QuorumInfo` instance but only by the `nodes()` which includes only controllers already in the quorum (the new ones are still observers).
If using the `describeCluster()`, it returns the brokers only instead.
This issue was also raised in this [discussion](https://lists.apache.org/thread/mq5n8j8w0ycg9fhd1m4zcc2ctrgfd0d4) on [KIP-1141](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=340037949).

> The suggestion here is to build the controller endpoint within the operator logic as it's already done within the `KafkaBrokerConfigurationBuilder`, because in the end it's the operator itself configuring the new controller so creating the advertised listeners list.

Reviewer 2: What does this mean? `KafkaBrokerConfigurationBuilder` builds a properties file. So I absolutely don't understand the relationship.

Author: Maybe this sentence is confusing but I am just trying to say that it will build the controller hostname (more than the complete endpoint) by using a similar logic we have within the `KafkaBrokerConfigurationBuilder` here https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java#L198 so my POC has something like this:

```java
RaftVoterEndpoint createControllerEndpoint(NodeRef controller) {
      String host = DnsNameGenerator.podDnsNameWithoutClusterDomain(
              reconciliation.namespace(),
              KafkaResources.brokersServiceName(reconciliation.name()),
              controller.podName());

      return new RaftVoterEndpoint(CONTROLPLANE_LISTENER_NAME, host, CONTROLPLANE_PORT);
}
```

There is no actual relationship with the `KafkaBrokerConfigurationBuilder`. It's just explaining that the logic to build the hostname is the same.

The operator then runs the registrations sequentially one by one, meaning it can only issue a new `addRaftVoter` request after the previous one completes.
Apache Kafka's internal logic only allows controllers to be promoted to voters once they are fully synchronized with the quorum.
When Apache Kafka receives a registration request, it writes a `VotersRecord` to the metadata log, and the controllers handle the quorum membership change.
This change can take time because the controller needs to catch up, and if the operator issues another `addRaftVoter` request within the same reconciliation, before the previous registration request completes, it will be refused.
In such cases, the operator stops further registrations, allowing the reconciliation to end, and the remaining registrations will proceed in subsequent reconciliations.
This means that while all controllers start up in a single reconciliation, registering them all as voters may require multiple reconciliation cycles.

Furthermore, monitoring that the controller has caught up with the active ones is not necessary.

> If the controller hasn't caught up yet and the operator runs the registration, the active controller returns an error and the reconciliation will fail, allowing the registration to be re-tried in the next reconciliation cycle.

Reviewer 2: That seems terrible. So the user will believe they have e.g. 7 controller nodes but only 3 will be registered because the rest did not caught up yet? I would expect Kafka to register all of them immediately.

Author: This is how Kafka works. It doesn't allow to register a controller if it's not caught up with the rest of the quorum it's trying to join. Am I missing anything here @fvaleri @mimaison @showuon ?

- Status: [{3:A}, {4:B}, {5:C}] (outdated, missing 0, 1, 2)

*Next reconciliation (automatic recovery):*
- Desired controllers: [0, 1, 2, 3, 4, 5]
- Status: [{3:A}, {4:B}, {5:C}] (outdated)
- DescribeQuorum: voters=[0:E, 3:A, 4:B, 5:C], observers=[1:F, 2:G]
- Analysis:
  - Node 0 in desired AND already in voters, no action needed
  - Nodes 1, 2 in desired but NOT in voters, need registration
- Operator registers 1:F and 2:G, **SUCCESS**

*Outcome (recovery):*
- DescribeQuorum: voters=[0:E, 1:F, 2:G, 3:A, 4:B, 5:C], observers=[]
- Status syncs from quorum: [{0:E}, {1:F}, {2:G}, {3:A}, {4:B}, {5:C}]

These scenarios demonstrate how the controller role label check ensures that only nodes that have been successfully rolled with the controller role are registered, and how the reconciliation process handles partial registration failures with automatic recovery.

### Metadata disk changes

When a controller's metadata disk is replaced or changed (e.g., PVC replacement, disaster recovery, JBOD disk configuration changes, moving metadata to a different disk), the controller will have a new directory ID because each disk has its own unique directory ID stored in the `meta.properties` file.

> The KRaft quorum reconciliation handles this scenario by detecting the directory ID mismatch and updating the quorum membership accordingly.

Reviewer 2: How do you deal with it when you do not know the new directory id yet etc?

Author: The (single controller) KRaft quorum reconciliation within the `KafkaRoller` happens after the node is rolled and the roller has detected it's ready. At that point, the new metadata disk has been formatted and the new directory ID is generated so it is read from the `meta.properties` via the `KafkaAgent` (this logic is still within the `KRaftQuorumReconciler` called by the roller).

#### Immediate handling during rolling updates

When a controller pod restarts and the metadata disk has changed, the `KafkaRoller` detects and handles this immediately through a "single-controller reconciliation":

* After the controller pod successfully restarts, the `KafkaRoller` invokes a "single-controller reconciliation" for that specific controller.
* The reconciliation queries the current quorum state using `describeMetadataQuorum()` and detects multiple incarnations of the same node ID:

>   * An old voter with the stale directory ID (from before the disk change).

Reviewer 2: We delete the metadata. So I assume the old voter will be defunct.

Author: It isn't actually how Kafka works. The old "incarnation" of the controller with the old directory ID needs to be unregistered to become an observer. And the new "incarnation" which starts up as observer needs to be registered to be promoted as voter.

  * Registers the new observer with the current directory ID using the `addRaftVoter` API.
* This happens immediately during the rolling update, minimizing the time the controller with a stale directory ID remains registered in the quorum.

The "single-controller reconciliation" follows the same two-phase approach as the full quorum reconciliation: unregistration first, then registration.
This ensures that even if the controller already exists as a voter with the old directory ID, it can be cleanly replaced.

#### Handling failures and recovery

If the single-controller reconciliation fails during the rolling update (e.g., unable to read directory ID from the pod, Kafka API errors, operator crash), the `KafkaRoller` will fail the rolling update for that controller:

* The rolling process throws an error, and the reconciliation is marked as failed.
* The controller remains running with the new directory ID but may still be registered with the old directory ID in the quorum, or not registered at all.
* No subsequent controllers in the rolling sequence are restarted until this issue is resolved.

> On the next reconciliation cycle, the full KRaft quorum reconciliation (at the end of the reconciliation) recovers from this failure:

Reviewer 4: Won't there be a full KRaft quorum reconciliation at the beginning of the reconciliation first?

Author: It's a typo there. It clearly refers to the KRaft quorum reconciliation at the "beginning" of the reconciliation. I will fix it.

*Outcome (recovery):*
- DescribeQuorum: voters=[3:A', 4:B', 5:C'], observers=[0:E, 1:F, 2:G, 3:A (garbage), 4:B (garbage), 5:C (garbage)]
- Status syncs from quorum: [{3:A'}, {4:B'}, {5:C'}] (status self-heals)

These scenarios demonstrate the two-phase execution (unregister before register) which prevents race conditions during disk changes, and how the full quorum reconciliation automatically recovers from failures during rolling updates by detecting directory ID mismatches and updating the quorum membership accordingly.

### Disaster recovery

Strimzi already provides a procedure for [recovering a deleted Kafka cluster](https://strimzi.io/docs/operators/latest/deploying#proc-cluster-recovery-volume-str).
It is related to the scenario where the PVs are still in place but all Kafka related custom resources together with the PVCs were deleted.
Because the PVs are still in place, all the metadata information is still available including the `meta.properties` file.

For clusters using dynamic quorum, the disaster recovery process requires additional steps to restore the controller directory IDs.

> Similar to how the `clusterId` must be manually retrieved through the PVCs and added to the `Kafka` custom resource status, users must also retrieve and provide the controller directory IDs.

Reviewer 2: As already raised above. We got a big pushback against the cluster ID. I cannot imagine us forcing a significantly worse version on the users.

Author: Let's keep the conversation where you already raised it.

If the operator successfully sets `kraft.version` to `1` but crashes or it is restarted before completing the migration (before updating the `controllers` field in the status or triggering the rolling update), the migration completes automatically on the next reconciliation.
This is possible because `kraft.version=1` is already persisted in Kafka.
When the operator restarts, it queries `kraft.version`, detects it is already set to `1`, and the `KRaftQuorumReconciler` proceeds to retrieve the controllers directory IDs.
The node configuration change to use `controller.quorum.bootstrap.servers` also triggers a rolling update of nodes and status update.
The idempotent nature of the KRaft Quorum Reconciler ensures that the migration completes safely regardless of when the operator restart occurred.

## Affected/not affected projects

Only the Strimzi Cluster Operator.

## Compatibility

When upgrading the Strimzi operator to a release supporting the dynamic quorum, existing clusters that use the static quorum will be automatically migrated to use dynamic quorum.
Of course, new Apache Kafka clusters are also deployed by using the dynamic quorum.

> ### Downgrade

Reviewer 2: I'm confused. You say that downgrade is not possible. But you pretend that it is fine. That seems inconsistent.

Author: I mean that downgrading to "static quorum" is not possible. When the cluster is migrated automatically to "dynamic quorum" but then you downgrade to the previous operator, the Kafka cluster will still continue to use dynamic quorum despite the nodes will be configured with a "voters" configuration because of course an old operator doesn't know about the "controllers bootstrap". But once the kraft.version is set to 1 (dynamic quorum) it can't come back to 0 (static quorum). So where do you see I was not clear in the description?

An alternative approach would be to simplify this logic by always using the `--initial-controllers` option with a dynamically generated controllers list.
On each reconciliation, the operator would build a complete list of all "desired" controllers (both existing and newly added) and use this list when formatting any controller's storage with the `--initial-controllers` option.
This would eliminate the need to differentiate between new cluster creation and scale-up scenarios in the formatting logic.

After investigation, this approach appears to work functionally.
A new controller formatted with `--initial-controllers` containing all current controllers can still join the quorum as an observer and be registered as a voter.
However, several issues make this approach unsuitable:

- Unnecessary checkpoint file: Formatting with `--initial-controllers` creates a bootstrap snapshot/checkpoint file on the scaled-up controller's disk. This file is redundant because new controllers don't use it to discover the quorum but they fetch the `VotersRecord` from the leader's metadata log instead. This can be considered a minor issue since the file is quite small and doesn't have any impact.

> - Split quorum risk: When multiple controllers are scaled up simultaneously and formatted with `--initial-controllers` including all newly added controllers, they can form a separate competing quorum independent of the existing voters if election timeouts expire before they fetch metadata from the active quorum. This violates Raft consensus safety guarantees by creating two independent quorums operating on the same cluster.

Reviewer 2: Can't this be avoided by scaling controllers always by one?

Author: Resolving this. Let's continue the discussion under Mickael comment https://github.com/strimzi/proposals/pull/203#discussion_r3016993034

> - Undocumented behavior: This approach is not documented in the official Apache Kafka documentation or KIP-853. Relying on undocumented behavior creates a risk that future Kafka versions could change or break this functionality without notice.

Reviewer 2: This seems like an interesting option. Should we document it then?

Author: Resolving this. Let's continue the discussion under Mickael comment https://github.com/strimzi/proposals/pull/203#discussion_r3016993034

> For these reasons, the proposal follows the official documented approach: using `--initial-controllers` only for initial cluster bootstrap and `--no-initial-controllers` for scale-up operations.

Reviewer 5: The reasons listed for discarding this option are pretty weak. 

- `Unnecessary checkpoint file`: This is pretty much irrelevant as the file is tiny.

- `Split quorum risk`: When formatting you should only add up to (n/2)-1 new controllers. That type of limitation is relatively common in distributed system so I would not consider it blocking.

- `Undocumented behavior`: As we discussed privately, this behavior works and we can document it in Apache Kafka if that would help.

Are there other technical reasons?

Reviewer 2: Yeah, that was pretty much my take as well here.

Author: > Unnecessary checkpoint file: This is pretty much irrelevant as the file is tiny.

I said this several times during my investigation during our discussions.

> Undocumented behavior: As we discussed privately, this behavior works and we can document it in Apache Kafka if that would help.

Tbh it's not exactly what we agreed. We had several discussions, I also sent an email about documenting this usage (no answer for a month, then only Luke replied), we also worked together on fixing a test and adding a new one about this scenario. But offline we seemed to agree with @showuon and his investigation so that my PR was closed and my email was answered. You can find references here:

https://lists.apache.org/thread/62p9svvdzgpgjd7o3kws3f917w6voqqn
https://github.com/apache/kafka/pull/21507

If things are different or you reached a different conclusion, I don't understand why not raising it during our previous discussions.

Also what does it mean for the Kafka project? Documenting that you can do it but with the limitation of adding up to (n/2)-1 only? It looks to me more stretching a way to do the formatting than documenting a working solution.

Reviewer 6: Mickael is correct, we can avoid the "split brain" issue by limiting the number to be scaled up/down one time. But I think that also need to add other logic to do the limitation and we have the same result here. So I think @ppatierno can add some more wordings here to emphasize the reason why you don't want to go this way. You must have some reason, like the UX is not good for users if users want to add the controller from 1 -> 3, which will be rejected, ... etc, to persuade readers.

Leveraging this feature was explored but the current semantic of auto-join doesn't work very well when scaling down a controller in a Kubernetes-based environment.
The above KIP states that, in order to unregister a controller when auto-join is enabled, you have to:

* shutdown the controller.
* run the unregistration command.

If you run the unregistration command as the first step while the controller is still running, it will join the quorum immediately because of the auto-join.

It doesn't work well in a scenario where the Apache Kafka cluster has combined-nodes and we want to scale down controllers by removing the "controller" role from one or more of them.
In this case, the node is not actually shut down and removed forever but it's rolled with a new configuration (i.e. the "controller" role is removed).
But in a Kubernetes-based environment where the rolling of a pod is driven by the platform, there is no opportunity to execute a step (like the controller unregistration) between the shutdown and startup.

> To overcome this issue, the best approach would be unregister the controller first and then rolling the node as broker only, but as mentioned before it would join again right after the unregistration.

Reviewer 2: Why can't you:
* Roll the pod to be in broker role only
* Unregister it from KRaft voters?

Given it is not controller anymore, it should not auto-join anymore, or? If it does, that seems like a major bug in Kafka.

Author: > If it does, that seems like a major bug in Kafka.

Kind of ... yes, here is what I opened https://issues.apache.org/jira/browse/KAFKA-19867
The discussion ended with the need of a re-design of auto-join which is not planned right now. @showuon can add more details to it.

* **Migration Barriers**: Organizations evaluating or migrating from ZooKeeper-based clusters to KRaft expect feature parity at minimum. The inability to scale controllers without downtime represents a regression that blocks adoption.
* **Incomplete KRaft Maturity**: While KRaft was introduced to eliminate ZooKeeper dependency and simplify Kafka's architecture, the lack of dynamic quorum support in Strimzi means users cannot fully leverage KRaft's capabilities as designed by the Apache Kafka project.

By implementing dynamic quorum support, Strimzi aligns with Apache Kafka's roadmap and enables users to operate KRaft-based clusters with the same (or better) operational flexibility they had with ZooKeeper, removing a major barrier to KRaft adoption in production environments.

## Proposal

This proposal is about supporting the dynamic controller quorum, thus the controllers scaling, within the Strimzi Cluster Operator, by going through three main steps:

* Add support for dynamic controller quorum to the Strimzi Cluster Operator and using it by default for any newly created Apache Kafka cluster.
* Add support for controllers scaling by leveraging the dynamic controller quorum.

> * Add migration from static to dynamic controller quorum.

Author: The first two bullets will come together for sure because strictly related.
The migration one can be done in a separate PR if we want. Meanwhile, the operator just skips the KRaft quorum reconciliation for any "static" quorum cluster (where the "controllers" field in the status is null).

> * Add migration from static to dynamic controller quorum.

Reviewer 3: The "Compatibility" section seems to imply that auto migration will be released at the same time.

During each reconciliation, the Strimzi Cluster Operator builds a `controllers` string from the controllers list in the status and passes it through the mounted ConfigMap to the `kafka_run.sh` script.
The string format is: `node-id@dns-name:port:directory-id,node-id@dns-name:port:directory-id,...`

Importantly, the operator only includes current controllers with the corresponding `directoryId` set when building this string.
This means that during scale-up operations, new controllers being added (which don't have a directoryId in the status yet) are excluded from the controllers string.
These new controllers will format their storage using the `-N` option and their directory IDs, automatically generated by Kafka, will be discovered later via the KRaft quorum reconciliation process (more details in the following sections).

The operator also adds a `cluster.new` field to the ConfigMap, which indicates whether this is a brand new cluster (`true`) or an existing cluster (`false`).
This field is determined based on the presence of the `clusterId` in the `Kafka` custom resource status:

> * A cluster is considered "new" if the status is null or the `clusterId` is null/empty.

Author: Referring to the `Kafka` custom resource `status`, when it's null, the cluster is new anyway. With regards to being static quorum based, it's the `controllers` field in the `status` to be null instead, but `status` still does exist.

Scaling controllers up or down within a KRaft quorum involves more than just spinning up new nodes or shutting down existing ones.
When a new controller is added to the Apache Kafka cluster, it has to be registered in order to join the KRaft quorum.
When a running controller is shut down, it has to be unregistered in order to leave the KRaft quorum.
The KRaft quorum reconciliation is a critical process that ensures the KRaft controller quorum membership accurately reflects the desired state of the cluster.
The main goal is reconciling the current KRaft quorum gathered from the Apache Kafka cluster with the desired controllers state requested by the users.
This process handles controller registration and unregistration needed for scaling up/down operations, metadata disk changes and disaster recovery.
It also takes part within the migration from static to dynamic quorum process.
The proposal is to have a new dedicated `KRaftQuorumReconciler` class to manage the necessary reconciliation steps.

The KRaft quorum reconciliation fits into the `KafkaReconciler` reconciliation process in this way:

* BEGIN RECONCILIATION
  * ... (other reconcile operations) ...

>   * **KRaft quorum reconciliation**: analyze current quorum state, unregister and register controllers as needed (typically unregisters controllers being scaled down).

Reviewer 6: OK, thanks.

  * scale down controllers.
  * ... (other reconcile operations) ...

>   * **KRaft quorum reconciliation** (rolling): for each controller pod restart, `KafkaRoller` invokes a "single-controller reconciliation" to handle metadata disk changes immediately (unregister old voter with stale directory ID, register new observer with current directory ID).

Author: Yes because it's reconcile the participation to the KRaft quorum only for that specific node and it's not a full KRaft quorum reconciliation which takes into account all controllers (the code in my POC is anyway shared between the two, because reconciling the KRaft quorum as a whole needs check quorum participation for each controller anyway).

>   * **KRaft quorum reconciliation** (rolling): for each controller pod restart, `KafkaRoller` invokes a "single-controller reconciliation" to handle metadata disk changes immediately (unregister old voter with stale directory ID, register new observer with current directory ID).

Reviewer 6: OK, thanks.

In the controllers scaling scenario, the operator handles the registration as well (not only the unregistration).

In addition to handling the registration and unregistration of controllers, the Strimzi Cluster Operator has to check the health of the KRaft quorum before allowing the scale down of the controllers.
At the beginning of the `KafkaReconciler` reconciliation process, if scaling down controllers could break the quorum (by losing the majority of voters needed for consensus), it should be blocked and reverted back by the operator.
A similar approach is already in place when attempting to scale down brokers that are hosting partitions, which could break HA.

#### How it works

The KRaft quorum reconciliation process follows these steps:

Query the current quorum state: Uses the Kafka Admin API `describeMetadataQuorum()` method to retrieve the current list of voters and observers, along with their directory IDs provided by the `QuorumInfo` object.

Analyze desired vs. actual state, for each "desired" controller node:
- Checks if the controller pod has been rolled with the controller role (verified by the presence of the `strimzi.io/controller-role` label on the pod).

> - Compares the controller's current state in the quorum (voter, observer, or absent) with the expected state based on the `controllers` status field.

Author: In this case the KRaft quorum reconciler logic won't do anything, nothing to register or nothing to unregister. On next reconciliation, it will still detect that 4th node is in the controllers "desired" list, gets the quorum metadata to compare and if it's an observer, it will be registered otherwise still skipped.

* by adding a new `KafkaNodePool` with `controller` role.
* by adding the `controller` role to an existing `KafkaNodePool` hosting `broker`-only nodes.

#### Scale-up scenarios

The following scenarios demonstrate how the KRaft quorum reconciliation handles controller scale-up operations, including both the success path and failure recovery mechanisms.

> Each node got from a describe quorum operation, both voter or observer, is represented as a couple `<nodeId>:<directoryId>` (i.e. `3:A`).

> Each controller within the `Kafka` custom resource status is represented with `{<nodeId>:<directoryId>}` (i.e. `{3:A}`) to reflect the JSON nature of the the `KafkaControllerStatus` object.

**Scenario 1.A: Scale up controller (success path)**

*Initial state:*
- Desired controllers (from spec): [3, 4, 5]
- Status: [{3:A}, {4:B}, {5:C}]
- DescribeQuorum: voters=[3:A, 4:B, 5:C], observers=[0:E, 1:F, 2:G]

*User action:*
- User scales up by adding controller 6
- Desired controllers become: [3, 4, 5, 6]

*Reconciliation:*
- Status before reconciliation: [{3:A}, {4:B}, {5:C}]

> - Analysis: Node 6 in desired but NOT in status, operator generates random UUID "D"

Author: It's a left over of the rejected approach, I will change it and review the rest if they have same issue.
FYI, it will be generated by the Kafka storage format tool when the node starts by using the famous `-N` option because it's a scale up.

* unregisters them from the quorum using the `removeRaftVoter` API, ensuring they are removed before the rolling update begins.

After unregistration completes:

* the operator builds the new node configuration without controller-specific settings.
* triggers a rolling update of the affected nodes.
* updates the `controller.quorum.bootstrap.servers` configuration on all remaining nodes to exclude the removed controllers.

During the rolling update, when each affected node restarts:

* the `kafka_run.sh` script detects the node is now a broker-only (no controller role in `process.roles`).
* the node starts without initializing controller storage, operating solely as a broker.
* the `strimzi.io/controller-role` label is removed from the pod.

The metadata stored on the controller's storage directory from when it was a combined node remains on disk but is no longer used or maintained.

If the reconciliation fails during the unregistration phase (e.g., Kafka returns an error), the rolling update will not proceed, preventing broker-only nodes from being restarted while still registered as voters in the quorum.

> #### Broker to combined node scenarios

Author: I wanted to have just one section to cover the combined node scenarios.

- Analysis: Nodes 0, 1, 2 in desired but NOT in voters, need registration
- Operator registers 0:E, **SUCCESS**
- Operator attempts to register 1:F, **FAILS**
- DescribeQuorum: voters=[0:E, 3:A, 4:B, 5:C], observers=[1:F, 2:G]
- Operator **CRASHES** before updating status

*Outcome (with crash):*
- Status: [{3:A}, {4:B}, {5:C}] (outdated, missing 0, 1, 2)

*Next reconciliation (automatic recovery):*
- Desired controllers: [0, 1, 2, 3, 4, 5]
- Status: [{3:A}, {4:B}, {5:C}] (outdated)
- DescribeQuorum: voters=[0:E, 3:A, 4:B, 5:C], observers=[1:F, 2:G]
- Analysis:
  - Node 0 in desired AND already in voters, no action needed
  - Nodes 1, 2 in desired but NOT in voters, need registration
- Operator registers 1:F and 2:G, **SUCCESS**

*Outcome (recovery):*
- DescribeQuorum: voters=[0:E, 1:F, 2:G, 3:A, 4:B, 5:C], observers=[]
- Status syncs from quorum: [{0:E}, {1:F}, {2:G}, {3:A}, {4:B}, {5:C}]

These scenarios demonstrate how the controller role label check ensures that only nodes that have been successfully rolled with the controller role are registered, and how the reconciliation process handles partial registration failures with automatic recovery.

> 

Reviewer 7: No example scenarios for the combined to broker-only node?

* Updates the `controllers` field in the `Kafka` custom resource status with the current directory IDs.

This recovery mechanism ensures that even if failures occur during rolling updates, the quorum membership will eventually converge to the correct state.
The idempotent nature of the reconciliation process means it can be safely retried across multiple reconciliation cycles until it succeeds.

The directory ID reading from the Kafka Agent is critical for this process.
If the agent is unavailable or the read fails, the reconciliation will fail and retry later, preventing incorrect assumptions about which directory ID is current.

#### Metadata disk change scenarios

The following scenarios demonstrate how the KRaft quorum reconciliation handles metadata disk changes during rolling updates, including both the success path and various failure recovery mechanisms.

**Scenario 4.A: Disk change, all operations succeed**

*Initial state:*
- Desired controllers (from spec): [3, 4, 5]
- Status: [{3:A}, {4:B}, {5:C}]
- DescribeQuorum: voters=[3:A, 4:B, 5:C], observers=[0:E, 1:F, 2:G]

*User action:*

> - Disk change on controller node pool (e.g., JBOD configuration change)

Reviewer 7: can you please explain the process that happens, after user edits the node pool. For example, before a rolling is triggered for a controller, is the directory id regenerated for it? Or does user specify the new directory id?

They mostly describe rejected approaches to solve specific problems within the overall dynamic quorum and controllers scaling support.

### Universal formatting via `--initial-controllers` option

The official Kafka documentation describes the differentiation between formatting initial controllers in a new cluster with the `-I` option, and formatting scaled-up controllers in an existing cluster with the `-N` option.

An alternative approach would be to simplify this logic by always using the `-I` option with a dynamically generated controllers list.
On each reconciliation, the operator would build a complete list of all "desired" controllers (both existing and newly added) and use this list when formatting any controller's storage with the `-I` option.
This would eliminate the need to differentiate between new cluster creation and scale-up scenarios in the formatting logic.

After investigation, this approach appears to work functionally.
A new controller formatted with `-I` containing all current controllers can still join the quorum as an observer and be registered as a voter.
However, several issues make this approach unsuitable:

> - Unnecessary checkpoint file: Formatting with `-I` creates a bootstrap snapshot/checkpoint file on the scaled-up controller's disk. This file is redundant because new controllers don't use it to discover the quorum but they fetch the `VotersRecord` from the leader's metadata log instead. This can be consider a minimal issue.

Reviewer 7: ```suggestion
- Unnecessary checkpoint file: Formatting with `-I` creates a bootstrap snapshot/checkpoint file on the scaled-up controller's disk. This file is redundant because new controllers don't use it to discover the quorum but they fetch the `VotersRecord` from the leader's metadata log instead. This can be considered a minor issue since the file is quite small and doesn't have any impact. 
```

> - Undocumented behavior: Most critically, this approach is not documented in the official Apache Kafka documentation or KIP-853. Relying on undocumented behavior creates a risk that future Kafka versions could change or break this functionality without notice.

Reviewer 6: LGTM. Thanks.
